### ISSUE TITLE
Readd an ability to define and to use custom Toolchain

### ISSUE THREAD

**Issue by ig-sinicyn** [PerfDotNet/BenchmarkDotNet#140]
Hi!

I continue my quest to enable BenchmarkDotNet to be used as an toolkit for performance unit tests. 
For this time, the biggest deal: performance.

As it is for now, it takes 30-80 seconds to run relatively simple benchmark. Also, the runner produces 50-100 mb of temp files. This is definitely a gamestopper.  
A typical perftests project can contain hundreds of perftests and waiting for a hour or two just to run them... well, this is bad:) 
Thing are getting much worse if the project is built using third-party CI service (such as appveyor.com).
Typical time limit on these rarely exceeds 30-40 minutes.

#### Now, the proposal: add an ability to run benchmarks in-process. 
There are no time spent to build the code, no temporary binaries. The speedup should be significant.  

In fact, it is. I've wrote fast-and-dirty proofcode... and it works:)

**DISCLAIMER:** [the code](https://github.com/ig-sinicyn/BenchmarkDotNet/commit/95fbef3d05c8366da89a50c22d72c560e46ef64f) is just a prototype, not a final solution. It definitely not complete, but it's good enough to proof that the idea works at all.

So, the results. Code from the develop branch, release build, an benchmark class with 20 benchmark methods.
* Classic toolchain: 35 seconds.
* Inline run toolchain: 1.2 seconds.

In both cases measurements are pretty close, difference is not bigger than between two consecutive runs under classic toolchain.

Now, the question: are you interested in it at all and does this proposal fit into the roadmap of the BenchmarkDotNet?

If yes, I'll need your guide to rewrite the current throw-it-out prototype into something more usable:)

**Comment by AndreyAkinshin** [comment]
Sounds interesting.

@mattwarren, @adamsitnik, what do you think?

**Comment by adamsitnik** [comment]
It would be really nice to have it as an option. I am not sure how much effort it would take to make Diagnosers compatible with this execution model.

**Comment by ig-sinicyn** [comment]
@AndreyAkinshin 
Ok. Some questions then:

1. What's the best way to plug it in? As far as I can see I should define a new Runtime enum member and add a new toolchain for it, right?
2. Is it safe to make [these members](https://github.com/ig-sinicyn/BenchmarkDotNet/commit/95fbef3d05c8366da89a50c22d72c560e46ef64f#diff-d727d4e1326fbe8a7d3152634926dd8f) of GeneratorBase virtual? Most of them should do nothing for in-process benchmarks
3. There's definitely work to be done with the counterpart of `BenchmarkProgram.txt` , `InlineGenerator.Program<T>`.
   There should be a way to keep these two in sync as much as it's possible, however it looks like there will be two versions of Program. First, non-generic, for void-returning benchmarks and second - for benchmarks that returns values. I do not want to emit the code as it makes much harder to keep it in sync with the text template file. Any suggestions for this are welcome! 
4. The worst part of the code is the InlineExecutor. It should be definitely rewritten from scratch, but I currently have no idea how to do it in the right way:)

And about diagnosers: what if we'll derive a new interface `IInProcessDiagnoser` with the same methods but with a `Thread` parameter, not a `Process`? InlineExecutor will throw if there will be any diagnosers not implementing `IInProcessDiagnoser`.

**Comment by mattwarren** [comment]
> It would be really nice to have it as an option. I am not sure how much effort it would take to make Diagnosers compatible with this execution model.

> And about diagnosers: what if we'll derive a new interface IInProcessDiagnoser with the same methods but with a Thread parameter, not a Process? InlineExecutor will throw if there will be any diagnosers not implementing IInProcessDiagnose

It may be that most/all Diagnosers can't run in-memory, But that's not a big problem. From what I understand this mode is for specific scenarion, if people want diagnostics they'll just have to run out-of-process.

**Comment by adamsitnik** [comment]
@ig-sinicyn 

> What's the best way to plug it in? As far as I can see I should define a new Runtime enum member and add a new toolchain for it, right?

Toolchain used to be configurable. But it was designed for the CoreClr support and once I implemented it I realised that it was not needed anymore, Runtime was the right setting. So I allowed myself to [remove it](https://github.com/PerfDotNet/BenchmarkDotNet/commit/de337116a0c1b84b90e20353b6a70eadba834f82) I believe that you should not extend Runtime, it is strictly for Clr type. I suggest Toolchain property ressurection.

> Is it safe to make these members of GeneratorBase virtual? Most of them should do nothing for in-process benchmarks

sure!

> There's definitely work to be done with the counterpart of BenchmarkProgram.txt , InlineGenerator.Program<T>. There should be a way to keep these two in sync as much as it's possible, however it looks like there will be two versions of Program. First, non-generic, for void-returning benchmarks and second - for benchmarks that returns values. I do not want to emit the code as it makes much harder to keep it in sync with the text template file. Any suggestions for this are welcome!
> The worst part of the code is the InlineExecutor. It should be definitely rewritten from scratch, but I currently have no idea how to do it in the right way:) 

I am not sure if current BenchmarkRunner and IToolchain suits your needs in the right way. If you find it not having too much sense to follow this pattern I would implement new runner, move the reused code to single class and reuse it in BenchmarkRunner. The other advantage would be less conflicts and separated responsibility. You could call it "InProcessBenchmarkRunner"

I am really sorry for giving this answer so late! I am also quite new person in the team, so don't take my suggestions as final! Please let me know if I could help you somehow.

**Comment by ig-sinicyn** [comment]
@adamsitnik, 
> If you find it not having too much sense to follow this pattern I would implement new runner, move the reused code to single class and reuse it in BenchmarkRunner. 

Well, I'd prefer to keep the current approach with the IToolchain. It allows to be as nonintrusive as it is possible and to implement my own toolchain in my project, doogfood it and only after that propose any changes to the Bench.Net.

So, the request must be changed to this:
**Please, add an ability to define and to use custom Toolchain.**

**Comment by adamsitnik** [comment]
@ig-sinicyn I will implement this for 0.9.6 release.

**Comment by ig-sinicyn** [comment]
@adamsitnik, Great news!
